python/ctsm/modify_fsurdat/modify_fsurdat.py

```diff
@@ -0,0 +1,288 @@
+            union_1 = np.logical_or(rectangle_1, rectangle_2)
+
+        if lat_1 <= lat_2:
+            # rectangles overlap
+            union_2 = np.logical_and(rectangle_3, rectangle_4)
+        else:
+            # rectangles don't overlap: one in the north, one in the south
+            union_2 = np.logical_or(rectangle_3, rectangle_4)
+
+        # union rectangles overlap
+        rectangle = np.logical_and(union_1, union_2)
+        not_rectangle = np.logical_not(rectangle)
+
+        return not_rectangle
+
+
+    def write_output(self, fsurdat_in, fsurdat_out):
+        """
+        Description
+        -----------
+        Write output file
+
+        Arguments
+        ---------
+        fsurdat_in:
+            (str) Command line entry of input surface dataset
+        fsurdat_out:
+            (str) Command line entry of output surface dataset
+        """
+
+        # update attributes
+        # TODO Better as dictionary?
+        title = 'Modified fsurdat file'
+        summary = 'Modified fsurdat file'
+        contact = 'N/A'
+        data_script = os.path.abspath(__file__) + " -- " + get_git_sha()
+        description = 'Modified this file: ' + fsurdat_in
+        update_metadata(self.file, title=title, summary=summary,
+                        contact=contact, data_script=data_script,
+                        description=description)
+
+        # abort if output file already exists
+        file_exists = os.path.exists(fsurdat_out)
+        if file_exists:
+            errmsg = 'Output file already exists: ' + fsurdat_out
+            abort(errmsg)
+
+        # mode 'w' overwrites file if it exists
+        self.file.to_netcdf(path=fsurdat_out, mode='w',
+                            format="NETCDF3_64BIT")
+        logger.info('Successfully created fsurdat_out: %s', fsurdat_out)
+        self.file.close()
+
+
+    def set_dom_nat_pft(self, dom_nat_pft, lai, sai, hgt_top, hgt_bot):
+        """
+        Description
+        -----------
+        In rectangle selected by user (or default -90 to 90 and 0 to 360),
+        replace fsurdat file's PCT_NAT_PFT with:
+        - 100 for dom_nat_pft selected by user
+        - 0 for all other non-crop PFTs
+        If user has specified lai, sai, hgt_top, hgt_bot, replace these with
+        values selected by the user for dom_nat_pft
+
+        Arguments
+        ---------
+        dom_nat_pft:
+            (int) User's entry of PFT to be set to 100% everywhere
+        lai:
+            (float) User's entry of MONTHLY_LAI for their dom_nat_pft
+        sai:
+            (float) User's entry of MONTHLY_SAI for their dom_nat_pft
+        hgt_top:
+            (float) User's entry of MONTHLY_HEIGHT_TOP for their dom_nat_pft
+        hgt_bot:
+            (float) User's entry of MONTHLY_HEIGHT_BOT for their dom_nat_pft
+        """
+
+        # initialize 3D variable
+        self.file['PCT_NAT_PFT'][:,:,:] = self.setvar(
+            self.file['PCT_NAT_PFT'][:,:,:], 0)
```

Comment:
- [ ] Are you sure this works correctly to set this 3-d variable? The mask (the not_rectangle) is defined in terms of a single 2-d slice, so I would think that it wouldn't work to use that mask directly to subset a variable with an extra dimension. But maybe there are some smarts in xarray to do this correctly? I just want to double-check: if you're not sure, it would be a good thing to play around with to make sure it's working right. (We can also do this together if you want.)

Reply:
Responded to this elsewhere.